### ISSUE TITLE
WorkflowTemplates are unable to reference global workflow (output) parameters/artifacts except as default values to inputs

### ISSUE THREAD

**Issue by samath117** [argoproj/argo#2590]
Checklist:

 * [x] I've included the version.
 * [x] I've included reproduction steps.
 * [x] I've included the workflow YAML.
 * [x] I've included the logs.

Global workflow parameters and WorkflowTemplates are almost entirely incompatible. When running `argo template create`, any global (output) parameters referenced cause the template linter to fail. Here's an example:
```
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: template-referencing-workflow-parameter
spec:
  templates:
  - name: print-name
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["{{workflow.parameters.message}}"]
```
When running `argo template create` on this file, I get this result:
```
Failed to create workflow template: templates.print-name templates.print-name: failed to resolve {{workflow.parameters.message}}
```

Instead, the template linter simply shouldn't complain about any references to the following:
```
{{workflow.parameters.X}}
{{workflow.outputs.parameters.X}}
{{workflow.artifacts.X}}
{{workflow.outputs.artifacts.X}}
```
Those could very well be defined earlier in the workflow that the template will be included in. A possible exception could be made for submittable templates, which should be validated as if they were a full workflow.

I mentioned a workaround, which I've been using so far:
```
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: template-referencing-workflow-parameter
spec:
  templates:
  - name: print-name
    inputs:
      parameters:
      - name: message
        value: "{{workflow.parameters.message}}"
    container:
      image: docker/whalesay:latest
      command: [cowsay]
      args: ["{{inputs.parameters.message}}"]
```

This template creates just fine and has the same behavior, as long as that input parameter is not overridden when the template is called. However, utilizing this notation adds endless unnecessary nesting and duplicated code.

At first, I buried this issue in the description to #2057, but it hasn't been resolved along with that one, so creating a separate issue. This issue persists in v2.7.0 despite the runtime resolution of parameters introduced in #2501.

---

**Message from the maintainers**:

If you are impacted by this bug please add a 👍 reaction to this issue! We often sort issues this way to know what to prioritize.

**Comment by simster7** [comment]
Closing as fixed in #2644

**Comment by kharmabum** [comment]
@simster7 we're running 3.1.8 and still seeing this issue.
